internal/pkg/log/log.go

```diff
@@ -2,105 +2,123 @@
+//
+// A self-contained usage example looks as follows:
+//
+//    Log.Named("foo-component").With(
+//        "field1", "value1",
+//        "field2", "value2",
+//     ).Error("my error message")
+type Logger interface {
+	// Named adds a sub-scope to the logger.
+	Named(name string) Logger
+
+	// With adds a variadic number of fields to the logging context.
+	// When processing pairs, the first element of the pair is used as the
+	// field key and the second as the field value.
+	//
+	// For example,
+	//
+	//   Log.With(
+	//     "hello", "world",
+	//     "failure", errors.New("oh no"),
+	//     "count", 42,
+	//     "user", User{Name: "alice"},
+	//  ).Info("this is an error message")
+	With(args ...interface{}) Logger
+
+	// Log a message at level Debug, annotated with fields provided through With().
+	Debug(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Info, annotated with fields provided through With().
+	Info(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Warn, annotated with fields provided through With().
+	Warn(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Error, annotated with fields provided through With().
+	Error(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Panic, annotated with fields provided through With(), and immediately
+	// panic.
+	Panic(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Fatal, annotated with fields provided through With(), and immediately
+	// call os.Exit.
+	Fatal(ctx context.Context, msg ...interface{})
+
+	// Sync flushes any buffered log entries.
+	Sync() error
 }
 
-// Opts allows manipulation of the Zap options.
-type Opts func(*Options)
+type logger struct {
+	zap *z.SugaredLogger
+}
 
-// addDefaults adds defaults to the Options
-func (o *Options) addDefaults() {
-	if o.OutputDest == nil {
-		o.OutputDest = os.Stderr
-	}
-	if o.Encoder == nil {
-		encCfg := zap.NewProductionEncoderConfig()
-		o.Encoder = zapcore.NewJSONEncoder(encCfg)
-	}
-	if o.Level == nil {
-		level := zap.NewAtomicLevelAt(zap.DebugLevel)
-		o.Level = &level
-	}
-	if o.StacktraceLevel == nil {
-		level := zap.NewAtomicLevelAt(zap.WarnLevel)
-		o.StacktraceLevel = &level
+// New returns an instance of Logger implemented using the Zap logging framework.
+func New(logLevel string) Logger {
+	zLevel, parseLogLevelErr := zap.ParseLogLevel(logLevel)
+
+	log := zap.New(
+		zap.Level(&zLevel),
+		// CallerSkip skips 1 number of callers, otherwise the file that gets
+		// logged will always be the wrapped file. In this case, log.go.
+		zap.AddCallerSkip(1),
+	)
+
+	if parseLogLevelErr != nil {
+		// Log an invalid log level error and set the default level to info.
+		log.Error("cannot set logger to desired log level")
 	}
+	return &logger{zap: log.Sugar()}
+}
 
-	o.ZapOptions = append(o.ZapOptions, zap.AddStacktrace(o.StacktraceLevel))
+func (l *logger) Named(name string) Logger {
```

Comment:
are we using a `Named` logger?

Reply:
This allows us to isolate components to do something like,
`logger.Named("mapper").Error(ctx, "failed to convert request")
with the output:
`{"level":"info","ts":1585006248.354409,"logger":"mapper","msg":"failed to convert request"}`